File: src/main/java/org/jlab/epsci/ersap/vtp/TwoStreamAggregator.java
package org.jlab.epsci.ersap.vtp;

import com.lmax.disruptor.*;

import static com.lmax.disruptor.RingBuffer.createSingleProducer;

public class TwoStreamAggregator {

    /**
     * VTP ports
     */
    private int vtpPort1;
    private int vtpPort2;

    /**
     * Max ring items
     */
    private final static int maxRingItems = 8192;

    /**
     * Ring buffers
     */
    private RingBuffer<RingRawEvent> ringBuffer1;
    private RingBuffer<RingRawEvent> ringBuffer2;
    private RingBuffer<RingRawEvent> ringBuffer12;

    /**
     * Sequences
     */
    private Sequence sequence1;
    private Sequence sequence2;
    private Sequence sequence12;

    /**
     * Sequence barriers
     */
    private SequenceBarrier sequenceBarrier1;
    private SequenceBarrier sequenceBarrier2;
    private SequenceBarrier sequenceBarrier12;

    private Receiver receiver1;
    private Receiver receiver2;
    private Aggregator aggregator12;
    private Consumer consumer;

    public TwoStreamAggregator(int vtpPort1, int vtpPort2) {
        this.vtpPort1 = vtpPort1;
        this.vtpPort2 = vtpPort2;

        ringBuffer1 = createSingleProducer(new RingRawEventFactory(), maxRingItems,
//                new LiteBlockingWaitStrategy());
                new YieldingWaitStrategy());
//                new SpinCountBackoffWaitStrategy(30000, new LiteBlockingWaitStrategy()));
        sequence1 = new Sequence(Sequencer.INITIAL_CURSOR_VALUE);
        sequenceBarrier1 = ringBuffer1.newBarrier();
        ringBuffer1.addGatingSequences(sequence1);

        ringBuffer2 = createSingleProducer(new RingRawEventFactory(), maxRingItems,
                new YieldingWaitStrategy());
        sequence2 = new Sequence(Sequencer.INITIAL_CURSOR_VALUE);
        sequenceBarrier2 = ringBuffer2.newBarrier();
        ringBuffer2.addGatingSequences(sequence2);


        ringBuffer12 = createSingleProducer(new RingRawEventFactory(), maxRingItems,
                new YieldingWaitStrategy());
        sequence12 = new Sequence(Sequencer.INITIAL_CURSOR_VALUE);
        sequenceBarrier12 = ringBuffer12.newBarrier();
        ringBuffer12.addGatingSequences(sequence12);

    }

    public void go() {
        receiver1 = new Receiver(vtpPort1, 1, ringBuffer1, 10);
        receiver2 = new Receiver(vtpPort2, 2, ringBuffer2, 10);

        aggregator12 = new Aggregator(ringBuffer1, ringBuffer2, sequence1,
                sequence2, sequenceBarrier1, sequenceBarrier2, ringBuffer12);

        int runNumber = 0;
        consumer = new Consumer(ringBuffer12, sequence12, sequenceBarrier12, runNumber);

        receiver1.start();
        receiver2.start();

        aggregator12.start();
        consumer.start();

    }

    public void close(){
        receiver1.exit();
        receiver2.exit();
        aggregator12.exit();
        consumer.exit();
    }

    public static void main(String[] args) {
        int port1 = Integer.parseInt(args[0]);
        int port2 = Integer.parseInt(args[1]);

        new TwoStreamAggregator(port1, port2).go();
    }

}

File: src/libsrc/TwoStreamAggregator.cpp


#include "TwoStreamAggregator.h"


namespace ersap {


    TwoStreamAggregator::TwoStreamAggregator(int vtpPort1, int vtpPort2) {
            this->vtpPort1 = vtpPort1;
            this->vtpPort2 = vtpPort2;

            // Try different wait strategies

//            uint32_t spinTries = 10000;
//            auto blockingStrategy = std::make_shared<Disruptor::BlockingWaitStrategy>();
//            auto waitStrategy = std::make_shared<Disruptor::SpinCountBackoffWaitStrategy>(spinTries, blockingStrategy);


            auto waitStrategy = std::make_shared<Disruptor::YieldingWaitStrategy>();
            ringBuffer1 = RingBuffer<std::shared_ptr<RingEvent>>::createSingleProducer(
                             RingEvent::eventFactory(), maxRingItems,  waitStrategy);
            sequence1 = std::make_shared<Disruptor::Sequence>(Disruptor::Sequence::InitialCursorValue);
            sequenceBarrier1 = ringBuffer1->newBarrier();
            std::vector<std::shared_ptr<Disruptor::ISequence>> writeSeqs1;
            writeSeqs1.push_back(sequence1);
            ringBuffer1->addGatingSequences(writeSeqs1);


            auto waitStrategy2 = std::make_shared<Disruptor::YieldingWaitStrategy>();
            ringBuffer2 = RingBuffer<std::shared_ptr<RingEvent>>::createSingleProducer(
                              RingEvent::eventFactory(), maxRingItems,  waitStrategy2);
            sequence2 = std::make_shared<Disruptor::Sequence>(Disruptor::Sequence::InitialCursorValue);
            sequenceBarrier2 = ringBuffer2->newBarrier();
            std::vector<std::shared_ptr<Disruptor::ISequence>> writeSeqs2;
            writeSeqs2.push_back(sequence2);
            ringBuffer2->addGatingSequences(writeSeqs2);


            auto waitStrategy12 = std::make_shared<Disruptor::YieldingWaitStrategy>();
            ringBuffer12 = RingBuffer<std::shared_ptr<RingEvent>>::createSingleProducer(
                               RingEvent::eventFactory(), maxRingItems,  waitStrategy12);
            sequence12 = std::make_shared<Disruptor::Sequence>(Disruptor::Sequence::InitialCursorValue);
            sequenceBarrier12 = ringBuffer12->newBarrier();
            std::vector<std::shared_ptr<Disruptor::ISequence>> writeSeqs12;
            writeSeqs12.push_back(sequence12);
            ringBuffer12->addGatingSequences(writeSeqs12);
        }

        void TwoStreamAggregator::go() {
            Receiver receiver1(vtpPort1, 1, ringBuffer1, 10);
            Receiver receiver2(vtpPort2, 2, ringBuffer2, 10);

            Aggregator aggregator12(ringBuffer1, ringBuffer2, sequence1,
                                    sequence2, sequenceBarrier1, sequenceBarrier2,
                                    ringBuffer12);

            int runNumber = 0;
            Consumer consumer(ringBuffer12, sequence12, sequenceBarrier12, runNumber);

            receiver1.startThread();
            receiver2.startThread();

            aggregator12.startThread();
            consumer.startThread();
            std::this_thread::sleep_for(std::chrono::seconds(4000));
        }

}

int main(int argc, char **argv) {
    char* p_end;

    int port1 = 45100, port2 = 45200;
    std::cout << "argc = " << argc << std::endl;

    if (argc > 1) {
        port1 = (int) strtol(argv[1], &p_end, 10);
    }

    if (argc > 2) {
        port2 = (int) strtol(argv[2], &p_end, 10);
    }

    std::cout << "port 1 = " << port1 << ", port 2 = " << port2 << std::endl;

    ersap::TwoStreamAggregator ag(port1, port2);
    ag.go();

    std::this_thread::sleep_for(std::chrono::seconds(4000));

    return 0;
}
